core/src/main/java/org/apache/iceberg/actions/BaseCommitService.java

```diff
@@ -0,0 +1,224 @@
+
+    completedRewrites = Queues.newConcurrentLinkedQueue();
+    committedRewrites = Lists.newArrayList();
+    inProgressCommits = Queues.newConcurrentLinkedQueue();
+  }
+
+  /**
+   * Perform a commit operation on the table for the set of file groups, should cleanup failed file
+   * groups.
+   *
+   * @param batch set of file groups
+   */
+  abstract void commitOrClean(Set<T> batch);
+
+  /**
+   * Clean up a specified file set by removing any files created for that operation, should not
+   * throw any exceptions
+   *
+   * @param group group of files which are not yet committed
+   */
+  abstract void abortFileGroup(T group);
+
+  /** Starts a single threaded executor service for handling file group commits. */
+  public void start() {
+    Preconditions.checkState(
+        running.compareAndSet(false, true), "Rewrite Commit service already started");
+    LOG.info("Starting commit service for {}", table);
+    // Partial progress commit service
+    committerService.execute(
+        () -> {
+          while (running.get() || completedRewrites.size() > 0 || inProgressCommits.size() > 0) {
+            try {
+              if (completedRewrites.size() == 0 && inProgressCommits.size() == 0) {
+                // Give other threads a chance to make progress
+                Thread.sleep(100);
+              }
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+              throw new RuntimeException("Interrupted while processing commits", e);
+            }
+
+            // commit whatever is left once done with writing.
+            if (!running.get() && completedRewrites.size() > 0) {
+              commitReadyCommitGroups();
+            }
+          }
+        });
+  }
+
+  /**
+   * Places a file group in the queue and commits a batch of file groups if {@link
+   * #rewritesPerCommit} number of file groups are present in the queue.
+   *
+   * @param group file group to eventually be committed
+   */
+  public void offer(T group) {
+    LOG.debug("Offered to commit service: {}", group);
+    Preconditions.checkState(
+        running.get(), "Cannot add rewrites to a service which has already been closed");
+    completedRewrites.add(group);
+    commitReadyCommitGroups();
+  }
+
+  /** Returns all File groups which have been committed */
+  public List<T> results() {
+    Preconditions.checkState(
+        committerService.isShutdown(),
+        "Cannot get results from a service which has not been closed");
+    return committedRewrites;
+  }
+
+  @Override
+  public void close() {
+    Preconditions.checkState(
+        running.compareAndSet(true, false), "Cannot close already closed RewriteService");
+    LOG.info("Closing commit service for {} waiting for all commits to finish", table);
+    committerService.shutdown();
+
+    boolean timeout = false;
+    try {
+      // All rewrites have completed and all new files have been created, we are now waiting for
```

Comment:
Can we fix the style of this comment? I think Spotless formatted the original comment and it looks weird now.

Reply:
Done